pkg/operators/wasm/datasource.go

```diff
@@ -0,0 +1,297 @@
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(ds))
+}
+
+// getDataSource returns a data source by its name.
+// Params:
+// - stack[0] is the name of the datasource (string encoded)
+// Return value:
+// - datasource handle
+func (i *wasmOperatorInstance) getDataSource(ctx context.Context, m wapi.Module, stack []uint64) {
+	dsName, err := stringFromStack(m, stack[0])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	ds := i.gadgetCtx.GetDataSources()[dsName]
+	if ds == nil {
+		i.logger.Debugf("datasource not found %q", dsName)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(ds))
+}
+
+// dataSourceGetField returns a handle to a data source.
+// Params:
+// - stack[0]: Datasource handle
+// - stack[1]: Field name
+// Return value:
+// - Field handle
+func (i *wasmOperatorInstance) dataSourceGetField(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("datasource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, stack[1])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc := ds.GetField(fieldName)
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+// dataSourceAddField add a field to the data source and returns its handle.
+// Params:
+// - stack[0]: Datasource handle
+// - stack[1]: Field name
+// - stack[2]: Field kind
+// Return value:
+// - Field handle
+func (i *wasmOperatorInstance) dataSourceAddField(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("datasource handle %d not found", stack[0])
+		stack[0] = 0
+		return
+	}
+	fieldName, err := stringFromStack(m, stack[1])
+	if err != nil {
+		i.logger.Warnf("reading string from stack: %v", err)
+		stack[0] = 0
+		return
+	}
+	acc, err := ds.AddField(fieldName, api.Kind(stack[2]))
+	if err != nil {
+		i.logger.Warnf("adding field %q to datasource %q: %v", fieldName, ds.Name(), err)
+		stack[0] = 0
+		return
+	}
+	stack[0] = wapi.EncodeU32(i.addHandle(acc))
+}
+
+// dataSourceSubscribe subscribes to the datasource.
+// Params:
+// - stack[0]: Datasource handle
+// - stack[1]: Priority
+// - stack[2]: Callback ID
+// Return value:
+// - 0 on success, 1 on error
+func (i *wasmOperatorInstance) dataSourceSubscribe(ctx context.Context, m wapi.Module, stack []uint64) {
+	ds, ok := i.getHandle(wapi.DecodeU32(stack[0])).(datasource.DataSource)
+	if !ok {
+		i.logger.Warnf("datasource handle %d not found", stack[0])
+		stack[0] = 1
+		return
+	}
+	prio := wapi.DecodeI32(stack[1])
+	cbID := stack[2]
+	cb := m.ExportedFunction("dsCallback")
+	if cb == nil {
+		i.logger.Warnf("module doesn't export dsCallback")
+		stack[0] = 1
+		return
+	}
```

Comment:
This could be checked on init and then the exported function cached on the `wasmOperatorInstance`.

Reply:
I cached the function on the instance, but I think the check needs to be done here. (That function is not mandatory in any case, right?)

Reply:
> That function is not mandatory in any case, right?

That's true if we don't define that as part of a "minimal to be implemented API" of guests. I'm fine either way.